api/kv.go

```diff
@@ -0,0 +1,328 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+// A KVClient is used to perform reads and writes against a KV secrets engine in Vault.
+//
+// The mount path is the location where the target KV secrets engine resides
+// in Vault. The version refers to the version of the target KV secrets engine.
+//
+// Vault development servers tend to have "secret" as the mount path
+// and 2 as the version, as these are the default settings when a server
+// is started in -dev mode.
+//
+// Learn more about the KV secrets engine here:
+// https://www.vaultproject.io/docs/secrets/kv
+type KVClient struct {
+	c         *Client
+	mountPath string
+	version   int
+}
+
+type KVSecret struct {
+	Data     map[string]interface{}
+	Metadata *VersionMetadata
+	Raw      *Secret
+}
+
+type KVMetadata struct {
+	CASRequired        bool                   `mapstructure:"cas_required"`
+	CreatedTime        time.Time              `mapstructure:"created_time"`
+	CurrentVersion     int                    `mapstructure:"current_version"`
+	CustomMetadata     map[string]interface{} `mapstructure:"custom_metadata"`
+	DeleteVersionAfter time.Duration          `mapstructure:"delete_version_after"`
+	MaxVersions        int                    `mapstructure:"max_versions"`
+	OldestVersion      int                    `mapstructure:"oldest_version"`
+	UpdatedTime        time.Time              `mapstructure:"updated_time"`
+	// Keys are stringified ints, e.g. "3"
+	Versions map[string]VersionMetadata `mapstructure:"versions"`
+}
+
+type VersionMetadata struct {
+	Version      int       `mapstructure:"version"`
+	CreatedTime  time.Time `mapstructure:"created_time"`
+	DeletionTime time.Time `mapstructure:"deletion_time"`
+	Destroyed    bool      `mapstructure:"destroyed"`
+	// There is currently no version-specific custom metadata.
+	// This field is just a copy of what's in the CustomMetadata field
+	// for the full KVMetadata of the secret.
+	CustomMetadata map[string]string `mapstructure:"custom_metadata"`
```

Comment:
Nice catch, looks like I had a change of heart on one of these but forgot to update the other! I think I will actually just change them both to be `map[string]string`, since elsewhere in the codebase people seem to be giving it that typing, and then users could do one less type assertion.

Reply:
Decided to go with map[string]interface{} after all as I was struggling with some type assertion stuff on our side, plus I think it's probably good to future-proof this...